### ISSUE TITLE
[DTensor] A New Placement -- RaggedShard

### ISSUE THREAD

**Issue by asd123www** [pytorch/pytorch#169320]
### 🚀 The feature, motivation and pitch

Hi PyTorch DTensor Team,

To make PyTorch DTensor even more powerful, we would like to gently propose adding a new DTensor placement called `RaggedShard`, which expresses *uneven sharding* across devices with *configurable sharding granularity* (e.g., non-shardable blocks). The existing Shard placement only supports even sharding and does not meet the needs of many modern applications, especially block-wise quantization and structure-aware training.

## Why
RaggedShard enables several important capabilities:
- *Block-wise quantization*: `RaggedShard` allows user-defined sharding block sizes, enabling communication-free quantization. In contrast, `Shard` cannot easily preserve block-wise quantization groups.
- *Structure-aware optimizers (e.g., Muon)*: `RaggedShard` supports gathering a full tensor onto a single device via the standard `dtensor.redistribute` interface, making these optimizers easier to use and more performant. In contrast, `Shard` cannot place a full tensor on a single device without overhead.
- *Zero-copy collectives in FSDP2*: `RaggedShard` eliminates FSDP2’s interleaved copies of Shard parameters around collectives, improving end-to-end training throughput by ~25% in production.

## How
The concept of `RaggedShard` can be illustrated as:

<img width="712" height="264" alt="Image" src="https://github.com/user-attachments/assets/4eecf061-bd9c-423a-8233-da76fd5f44d2" />

`RaggedShard` is defined by two parameters, `dims` and `local_units`, which allow it to express element-wise, symmetric/asymmetric row-wise, and block-wise sharding granularities:

```python
class RaggedShard(Placement):
    """
    The ``RaggedShard`` placement specifies DTensor sharding based on tensor
    memory continuity.

    Args:
        dims (tuple[int]): The prefix dimension(s) to shard.
            For example, if a tensor of shape (n, m, k) is ragged-sharded with
            dims=(0, 1), then each ``local_tensor.numel()`` is guaranteed to be
            a multiple of k.
        local_units (tuple[int]): A tuple of integers specifying the relative
            allocation of elements across shards. The length of the tuple must
            equal ``mesh.size(dim)``.

            For example, ``local_units=(1, 2, 1, 1)`` assigns:
            - 20% of elements to device 0
            - 40% to device 1
            - 20% each to devices 2 and 3
    """

    dims: tuple[int, ...]
    local_units: tuple[int, ...]
```

The full design of `RaggedShard` can be found [here](https://github.com/volcengine/veScale/blob/main/docs/texts/raggedshard.md). It has been deployed in large-scale production for a long time, enabling numerous internal training workloads and providing both flexibility and performance.

## Hope

We sincerely hope the PyTorch DTensor team will consider `RaggedShard` as a new DTensor placement. Alternatively, we would welcome a simpler interface for registering custom DTensor placements *without* intrusive modifications to DTensor internals (especially dispatch and redistribution). Such an interface would make it much easier for the community to innovate and iterate on DTensor features, making PyTorch DTensor great!

@leonardo0lyj @jc-bytedance @pengyanghua

**Comment by asd123www** [comment]
@wconstab, Great proposal! @ezyang, Great blog! Our veScale team have followed your link and carefully studied the `ShardingPlaceholder`. We have several thoughts below:

1.  **Nature of the Placeholder**: `ShardingPlaceholder` itself seems to function as a *dynamic placement* mechanism (i.e., a placement placeholder that expands to either `Shard` or `StridedShard` during runtime [[code](https://github.com/pytorch/pytorch/pull/167677/files#diff-ef04401bd225f19f02d60795294e50fd58f1bc49d9653a8b0fddb0bf00609c89R65)]), rather than a fully custom placement.

2.  **Extensibility**: Fortunately, the redesigned sharding propagation path of `ShardingPlaceholder` could be extended for `RaggedShard` or other custom placements, especially since it creates an isolated branch [[code](https://github.com/pytorch/pytorch/pull/167677/files#diff-f2022847de0552d77a9a79764ec095614099944e28a5ee82829b11c6c4ba4213R409-R482)] before entering the regular sharding prop using `Shard`/`Replicate`/`Partial`.

3.  **Implementation Plan**: The full picture for registering a custom placement like `RaggedShard` might look like this:
    * A **definition** like `class CustomPlacement(Placement)`, which can be subclassed to `RaggedShard` or `PartitionShard` or future placement.
    * `CustomPlacement` has an `__init__` method to be overridden by subclasses, allowing definitions via arbitrary meta info (including 'symbolic ints').
    * **For computation**: `CustomPlacement` goes through an isolated branch of sharding prop (like `ShardingPlaceholder`) in `_propagate_tensor_meta_non_cached` (maybe via a hook) and lets subclasses register their custom per-operator strategies (or even run custom kernel like online-softmax).
    * **For communication**: `CustomPlacement` goes through an isolated branch of `redistribute_local_tensor` (maybe via a hook) before entering the 'three-placement redistribute' logic (i.e., `Shard`/`Replicate`/`Partial` -> `Shard`/`Replicate`/`Partial`). This allows subclasses to register their custom redistribute logic, such as `RaggedShard` to `Replicate` [[code1](https://github.com/volcengine/veScale/blob/cb8ba57b178619813f58fbba17bc8048f4c7dbb8/vescale/dtensor/placement_types.py#L128)] [[code2](https://github.com/volcengine/veScale/blob/cb8ba57b178619813f58fbba17bc8048f4c7dbb8/vescale/dtensor/_redistribute.py#L100)].
    * **For DCP**: `CustomPlacement` can also go through an isolated branch for DCP write/read within `DTensor` method [[code](https://github.com/volcengine/veScale/blob/cb8ba57b178619813f58fbba17bc8048f4c7dbb8/vescale/dtensor/_api.py#L550)].
    * **For Strided CustomPlacement**: intuitively, using ordered shard can avoid the complexity of creating another strided version of custom shard; however, if supporting ordered shard is too hard, the strided `CustomPlacement` like `StridedRaggedShard` can be also be implemented as a companion placement, which should go into the branch of `ShardingPlaceholder` for dynamic replacement. (The catch is that strided custom placement can still be complex when composing with other strided placement.😅)
    
    
 cc [@leonardo0lyj](https://github.com/leonardo0lyj) [@jc-bytedance](https://github.com/jc-bytedance)
